### DISCUSSION TITLE
Help with Installing on Digital Ocean

### DISCUSSION THREAD

**Discussion by osseonews** [pocketbase/pocketbase#512] [Q&A]
I'm a newb with this, so the questions below might seem silly to advanced users, but I'd really appreciate some help setting this up on Digital Ocean. I am extremely impressed with this software, and I use/tried every single other type of nocode database software like this, and frankly pocketbase blew me away (I can detail reasons in another post). So I'd really like to start testing it. Anyway, below are the steps I figure are needed to get this running on a DO droplet with questions. BTW, I usually just use DO's marketplace applications for software I use, so I never really set up much myself, except docker packages.

- [ ] Spin up a droplet: Which DO droplet will work for this? Just their standard ubuntu?
- [ ] Download and extract the linux zip.  Which one? There are 3. AMD64, ARM64, ARMV7? 
- [ ] Upload the pocketbase binary to droplet. Well a few questions here: What is the binary file? It is an .exe file? And then how do you upload this file to DO. I assume this is done with some sort of command line. If so, what is the command to use? Which folder do we upload it into in DO?
- [ ] Create a systemd service to keep the process up. This I have no clue at all about. How do I get a systemd service started on DO? What are the CLI commands? And then what are the exact commands to keep the process up?
- [ ] Assign a readable URL to the service. I assume once the service is running you use at the droplet IP address with the port. But what port is it running at? And then if we want to run it on a subdomain, i.e. pocketbase.domain.com - how do we set this up? Assume we need Nginx installed? Anyone have a good guide for that? I did set up Nginx with SSL on DO a few years back, and got it to work, but boy was that hell for me and unfortunately, I didn't document the steps I took to get it working, so now I have no clue again.
- [ ] Getting and Downloading the database: After you get everything up and running, it is obviously critical to make sure you don't lose all your data suddenly, if the droplet fails for some reason or some other thing happens. So how would you go about locating or downloading the database? Is there any command to upload the database to a DO spaces on a daily basis as a backup?

Thanks for your help.

**Comment by ganigeorgiev** [comment]
> Spin up a droplet: Which DO droplet will work for this? Just their standard ubuntu?

Any DO droplet will work (even the cheapest $4 offering). Yes, standard Ubuntu is fine.

---

> Download and extract the linux zip. Which one? There are 3. AMD64, ARM64, ARMV7?

If I'm not mistaken, DO droplets uses x86_x64 architecture so you can download the latest linux amd64 bundle - https://github.com/pocketbase/pocketbase/releases/download/v0.8.0/pocketbase_0.8.0_linux_amd64.zip

---

> Upload the pocketbase binary to droplet. Well a few questions here: What is the binary file? It is an .exe file? And then how do you upload this file to DO. I assume this is done with some sort of command line. If so, what is the command to use? Which folder do we upload it into in DO?

Once you've downloaded and extracted the zip, you'll find 2 files - `LICENSE` and `pocketbase`. The executable/binary is `pocketbase`. You need to upload only the `pocketbase` file to the droplet. There are several ways to do this. Personally I prefer [rsync](https://linux.die.net/man/1/rsync):

```
rsync -avz -e ssh /local/path/to/pocketbase root@YOUR_DROPLET_IP:/root/pb
```

This will upload the `pocketbase` binary into the `/root/pb` directory in the droplet.

---

> Create a systemd service to keep the process up. This I have no clue at all about. How do I get a systemd service started on DO? What are the CLI commands? And then what are the exact commands to keep the process up?

_(The below examples are using `vi` as text editor, but if you prefer you can use `nano` I'm just not sure if it is installed by default)_

Run `vi /lib/systemd/system/pocketbase.service`.

Press `CTRL + I` to enter insert mode and paste the following content:

```nginx
[Unit]
Description = pocketbase

[Service]
Type           = simple
User           = root
Group          = root
LimitNOFILE    = 4096
Restart        = always
RestartSec     = 5s
StandardOutput = append:/root/pb/errors.log
StandardError  = append:/root/pb/errors.log
ExecStart      = /root/pb/pocketbase serve --http="yourdomain.com:80" --https="yourdomain.com:443"

[Install]
WantedBy = multi-user.target
```
(**replace _yourdomain.com_ with your actual domain/subdomain**).

And then press `ESC` and type `:wq` to save the file.


To enable and start the service, run `systemctl enable pocketbase.service --now`

---

> Assign a readable URL to the service. I assume once the service is running you use at the droplet IP address with the port. But what port is it running at? And then if we want to run it on a subdomain, i.e. pocketbase.domain.com - how do we set this up? Assume we need Nginx installed? Anyone have a good guide for that? I did set up Nginx with SSL on DO a few years back, and got it to work, but boy was that hell for me and unfortunately, I didn't document the steps I took to get it working, so now I have no clue again.

You don't need nginx or apache. PocketBase comes with builtin server and Let's encrypt generation and everything will be handled by the above systemd service. You need only to assign a domain/subdomain to your droplet. You can follow this DO guide - https://docs.digitalocean.com/products/networking/dns/how-to/add-domains/.

---

> Getting and Downloading the database: After you get everything up and running, it is obviously critical to make sure you don't lose all your data suddenly, if the droplet fails for some reason or some other thing happens. So how would you go about locating or downloading the database? Is there any command to upload the database to a DO spaces on a daily basis as a backup?

Your PocketBase files will be stored by default `pb_data` directory that will be created automatically when you first start the application. Following the above example, the full path to the directory would be `/root/pb/pb_data/`.

There is no integrated command to copy periodically your `pb_data` directory to a DO spaces, but 3rd party tools exist, like [rclone](https://github.com/rclone/rclone) that will do that for you if combined with a cron job. Alternatively there is also [Litestream](https://litestream.io/).

---

**Footnotes**
I recommend you to first get familiar with PocketBase locally before directly deploying to production because it may turn out that it may not be suitable for your use case. To do that, download the executable for your local platform and then run:
```
./pocketbase serve
```
The above will start a local http server on `127.0.0.1:8090` that you can explore and experiment with.

**Reply by paulsutherland** [reply]
@osseonews @ganigeorgiev 

Whoops sorry I did not realise you answered this and I expanded on the previous discussion here. #480.

Cheers...
Paul

**Reply by heyarviind** [reply]
How to serve nextjs with pocketbase on digitalocean?

**Reply by ganigeorgiev** [reply]
@heyarviind That depends on how you use nextjs -  Static Site Generation/Client-Side Rendering, Server-side Rendering or mixed. For the first option you can just ship the generated static files and put them in your `pb_public` directory. For the other two - you'll need to install nodejs and use nginx/apache/caddy/etc. as a reverse proxy, because now you'll have 2 servers running (node and PocketBase).

**Reply by heyarviind** [reply]
@ganigeorgiev thank you for the reply, one last question.

I have hosted Pocketbase in the Digitalocean server located in Bangalore and attached a volume to store the data in the same location. When ever I access the Pocketbase API it takes more than 800 ms to fetch and return the data (Even I am in same location).

 Is there any way to make it faster and reduce the latency?

**Reply by ganigeorgiev** [reply]
It's hard to tell because a lot of things could contribute to slow response times - your network connection, the DO region network infrastructure, the available resources of your droplet, whether you use a VPN, etc.

_If you can reproduce the slow response times locally with the same db, then it is also possible that some particular PocketBase query is not optimized. If that's the case, please open another discussion with steps how to reproduce it and I'll investigate it in more details._

**Comment by osseonews** [comment]
Thank you! Will try to install based on the above.

**Reply by ganigeorgiev** [reply]
@osseonews Just to let you know that now there is quick start guide how to deploy on Fly.io for free - https://github.com/pocketbase/pocketbase/discussions/537

**Comment by osseonews** [comment]
BTW, when try set up UFW on Digital ocean with the command, `ufw allow 80,443 `you get an error: "Must specify 'tcp' or 'udp' with multiple ports". Do you know which to use with Pocketbase? tcp or udp?

**Reply by ganigeorgiev** [reply]
tcp, but on DO this ports are usually open by default I think. You can run `ufw status` to verify it. I'll update the command in the docs.

**Reply by tigawanna** [reply]
Obviously don't forget to enable port 22 for ssh before exiting

**Comment by chopfitzroy** [comment]
Hey there, I have automated @ganigeorgiev instructions above into an Ansible playbook [here](https://github.com/chopfitzroy/api.coffeeandcode.app) for anyone interested.

The only problem I am having is I am getting the following error in `errors.log` when trying to access `api.coffeeandcode.app/_/`:

```
2022/12/28 03:08:41 http: TLS handshake error from 192.241.196.4:39796: 400 urn:ietf:params:acme:error:accountDoesNotExist: Account "https://acme
-v02.api.letsencrypt.org/acme/acct/890828137" not found
2022/12/28 03:08:41 http: TLS handshake error from 192.241.196.4:40700: acme/autocert: missing certificate
2022/12/28 03:08:41 http: TLS handshake error from 192.241.196.4:40712: acme/autocert: missing certificate
2022/12/28 03:08:41 http: TLS handshake error from 192.241.196.4:40714: tls: client requested unsupported application protocols ([http/0.9 http/1
.0 spdy/1 spdy/2 spdy/3 h2c hq])
2022/12/28 03:08:41 http: TLS handshake error from 192.241.196.4:40726: tls: client requested unsupported application protocols ([hq h2c spdy/3 s
pdy/2 spdy/1 http/1.0 http/0.9])
2022/12/28 03:08:41 http: TLS handshake error from 192.241.196.4:40730: acme/autocert: missing certificate
2022/12/28 03:08:41 http: TLS handshake error from 192.241.196.4:40738: acme/autocert: missing certificate
2022/12/28 03:08:41 http: TLS handshake error from 192.241.196.4:40744: acme/autocert: missing certificate
2022/12/28 03:08:41 http: TLS handshake error from 192.241.196.4:40748: acme/autocert: missing certificate
2022/12/28 03:08:41 http: TLS handshake error from 192.241.196.4:40750: acme/autocert: missing certificate
  ```
  
 I am not sure if this is related to my [DNS config](https://www.digitalocean.com/community/tools/dns?domain=api.coffeeandcode.app), my DO droplet, or how PocketBase registers SSL certs, any help would be much appreciated.

**Reply by ganigeorgiev** [reply]
If you are using the IP when you starting PocketBase with `--https`, then it won't work. This is because Let's encrypt doesn't allow issuing certificates on bare IPs and you'll need a valid public domain name, aka:

```sh
./pocketbase serve --http="yourdomain.com:80" --https="yourdomain.com:443"
```

**Reply by chopfitzroy** [reply]
Oops, my apologies when I originally tried that I was getting an error along the lines of `dial tcp: lookup 67.207.67.2 no such host` which appears to have disappeared now so I would assume that is related to propagation.

When using:
```
./pocketbase serve --http="api.coffeeandcode.app:80" --https="api.coffeeandcode.app:443"
```

I am seeing the following output in `errors.log`:
```
2022/12/28 07:43:57 http: TLS handshake error from 151.210.149.155:38965: remote error: tls: unknown certificate
2022/12/28 07:43:57 http: TLS handshake error from 151.210.149.155:39029: remote error: tls: unknown certificate
2022/12/28 07:45:57 http: TLS handshake error from 209.141.55.120:56920: acme/autocert: host "164-92-117-27.luminatinet.com" not configured in Ho
stWhitelist
2022/12/28 07:45:57 http: TLS handshake error from 209.141.33.65:46130: acme/autocert: missing server name
```

**Reply by ganigeorgiev** [reply]
@chopfitzroy Sending a curl request to `https://api.coffeeandcode.app` seems to resolve successfully so I guess the above error logs are from bots trying to access your server via its IP/some other domain or doesn't follow the http redirect, I'm not really sure (_`remote errors: ...` are usually error response from the clients, aka. for some reason the other side of the connection was not able/didn't want to verify the certificate_).

**Reply by chopfitzroy** [reply]
Interesting... I just tried accessing it through the browser and it all works now? I wonder if again it could have been some propagation issue?

Regardless thank you!
